src/functions.php

```diff
@@ -12,22 +12,29 @@
  * Creates a `Promise` which resolves with the stream data buffer
  *
  * @param ReadableStreamInterface $stream
+ * @param int $maxLength
  * @return Promise\CancellablePromiseInterface Promise<string, Exception>
  */
-function buffer(ReadableStreamInterface $stream)
+function buffer(ReadableStreamInterface $stream, $maxLength = null)
 {
     // stream already ended => resolve with empty buffer
     if (!$stream->isReadable()) {
         return Promise\resolve('');
     }
 
+    $deferred = new Promise\Deferred();
     $buffer = '';
-    $bufferer = function ($data) use (&$buffer) {
+    $bufferer = function ($data) use (&$buffer, $deferred, $maxLength) {
         $buffer .= $data;
+        if ($maxLength !== null && isset($buffer[$maxLength])) {
+            $deferred->reject(new \RuntimeException('Buffer exceeded maximum length'));
```

Comment:
I would suggest using an `OverflowException` here and explicitly document this in the README. This way, the user can easily discern normal stream errors from buffer overflow errors :+1:

Reply:
Same as my comment here https://github.com/reactphp/http/issues/232#issuecomment-336187806, `OverflowException` would be nice.